src/can.rs

```diff
@@ -0,0 +1,357 @@
+                tseg_2: 4,
+                sjw: 3,
+                triple_sampling: false,
+            },
+            Timing::B800K => twai_timing_config_t {
+                brp: 4,
+                tseg_1: 16,
+                tseg_2: 8,
+                sjw: 3,
+                triple_sampling: false,
+            },
+            Timing::B1M => twai_timing_config_t {
+                brp: 4,
+                tseg_1: 15,
+                tseg_2: 4,
+                sjw: 3,
+                triple_sampling: false,
+            },
+        }
+    }
+}
+
+/// Is used to filter out unwanted CAN IDs (messages).
+///
+/// Notice that Espressif TWAI (CAN in rest of the world) acceptance filtering
+/// works differently than common CAN filtering (for example mask bits are inversed).
+/// However here those differences are hidden away from the user and common CAN filtering can be used.
+///
+/// `mask` is used to determine which bits in the CAN ID are compared with the `filter`.
+/// If a mask bit is set to a zero, the corresponding CAN ID bit will be accepted,
+/// regardless of the value of the filter bit.
+///
+/// `filter` determines which bits must be set in the CAN ID to accept the frame,
+/// however `mask` can relax those constraints if its bit is set to zero.
+///
+/// ## Examples
+///
+/// This shows how 11 bit CAN ID `0x3AA` goes through filtering engine and is finally accepted:
+/// ```
+/// // can id    [ 0 1 1 1 0 1 0 1 0 1 0 ]
+/// // mask      [ 1 0 1 0 0 1 1 1 0 0 0 ]
+/// //             1 = compare
+/// //             0 = do not care
+/// // masked id [ 0 _ 1 _ _ 1 0 1 _ _ _ ]
+/// // filter    [ 0 0 1 1 1 1 0 1 0 1 1 ]
+///
+/// // can id    [ 0 1 1 1 0 1 0 1 0 1 0 ]
+/// // accepted
+/// ```
+///
+/// Notice that for example `0x7AA` would not be accepted because its MSB bit is `1`,
+/// but `filter` only accepts `0` in this bit position and `mask` says that this bit must be compared.
+///
+/// Accept only CAN ID `0x567`
+/// ```
+/// let filter = 0x567;
+/// // every bit must match filter
+/// let mask   = 0x7FF;
+/// let f = Filter::Standard { filter, mask };
+/// ```
+///
+/// Accept CAN IDs `0x560 - 0x56F`
+/// ```
+/// let filter = 0x560;
+/// // do not care about 4 LSB bits
+/// let mask   = 0x7F0;
+/// let f = Filter::Standard { filter, mask };
+/// ```
+pub enum Filter {
+    // Filter for 11 bit standard CAN IDs
+    Standard { filter: u16, mask: u16 },
+    // Filter for 29 bit extended CAN IDs
+    Extended { filter: u32, mask: u32 },
+}
+
+impl Filter {
+    /// Filter that allows all standard CAN IDs.
+    pub fn standard_allow_all() -> Self {
+        Self::Standard {
+            filter: 0,
+            mask: 0x7FF,
+        }
+    }
+
+    /// Filter that accepts all extended CAN IDs.
+    pub fn extended_allow_all() -> Self {
+        Self::Extended {
+            filter: 0,
+            mask: 0x1FFFFFFF,
+        }
+    }
+}
+
+/// CAN abstraction
+pub struct CanBus<TX: OutputPin, RX: InputPin> {
```

Comment:
The way `CanBus` is modeled currently, there is nothing stopping me from creating multiple instances of `CanBus`.
However - by looking at the ESP-IDF API, I don't think multiple controllers are supported - the TWAI controller seems to be a singleton instance. E.g. all of the CAN Bus methods are globals that do not take a "can bus"/"controller" parameter or anything like that (see `twai_start`, `twai_transmit` and so on).

To remedy this, I suggest we do the following (basically the same pattern that is followed in the ADC, I2C, SPI and UART drivers):
* Define a struct, call it `CAN` or `TWAI` similar to [this one](https://github.com/esp-rs/esp-idf-hal/blob/master/src/adc.rs#L311). (You don't need to do it with a macro, because all esp32* chips seem to contain a single TWAI controller.)
* The struct has to have this strange `PhantomData` member, and should have a `new` method which is marked with `unsafe`, so that the user is not tempted to call `new` on it directly. Remember - this is a singleton.
* The struct also needs to be [`Send` so that we can send it across threads](https://github.com/esp-rs/esp-idf-hal/blob/master/src/adc.rs#L322)
* The struct does not need to implement a `Can`/`Twai` trait or anything like that as the `ADC1` and `ADC2` structs do, because we have "a" single Can/Twai controller, so no need for additional identification of it in front of ESP-IDF
* Define a `can`/`twai` member of type `CAN`/`TWAI` in [peripherals.rs](https://github.com/esp-rs/esp-idf-hal/blob/master/src/peripherals.rs#L39) and [instantiate it](https://github.com/esp-rs/esp-idf-hal/blob/master/src/peripherals.rs#L82) using the `unsafe new` method.
* Modify your `CanBus`/`TwaiBus`'s `new` to take an additional parameter: a struct of type `CAN`/`TWAI` and keep this instance inside the `CanBus` instance. Since the `CAN` struct has only a single `PhantomData` member, the compiler will optimize away all its usages when generating machine code. These equilibristics are only to help the user understand and follow the contract that esp32*s have a single CAN controller.
* Implement a `CanBus::release(self) -> (CAN, TX, RX)` method, that calls `twai_driver_uninstall` (or whatever necessary) to destruct the `CanBus` instance and returns the `CAN` singleton instance as well as the pins that were injected into the `CanBus` instance during its construction.